### ISSUE TITLE
WindowPoStScheduler setSender has a logic error

### ISSUE THREAD

**Issue by GFZRZK** [filecoin-project/lotus#5480]
**Describe the bug**
Suppose we only use worker address to sign block info and left zero fil in it (since we can seal & submit wdpost message with controller address), it will return a no balance error in [gm, err := s.api.GasEstimateMessageGas(ctx, msg, spec, types.EmptyTSK)](https://github.com/filecoin-project/lotus/blob/7e55c44d4e09a058782e07cf172c96fb5546b172/storage/wdpost_run.go#L777), and it returns with no selecting on address, though we have enough fil in control address. It's not a good fallback and should change some logic here.

**Comment by jennijuju** [comment]
@GFZRZK could you please add an expected behaviour you want to suggest?

**Comment by GFZRZK** [comment]
maybe just follow the order control -> owner -> worker is better? Then we just need to focus on the control address's balance. Anyway worker address is a fallback option as comments said, so it's confused when the control address has enough fil but setSender return an error

**Comment by jennijuju** [comment]
> maybe just follow the order control -> owner -> worker is better? Then we just need to focus on the control address's balance. Anyway worker address is a fallback option as comments said, so it's confused when the control address has enough fil but setSender return an error

Can you confirm your node was in sync and control address has enough ⨎? B/c like you said, it’s currently using control address when it’s available, then fail to owner and eventually worker.

**Comment by GFZRZK** [comment]
Yes, I'm sure my control address has enough fil, and you can review the code of the func I posted above, it uses the `s.api.GasEstimateMessageGas `, at that time, the msg's from is worker address, and it will get an `actor balance less than needed` error, then the func just return and will not check whether any other address is available

**Comment by jennijuju** [comment]
> Yes, I'm sure my control address has enough fil, and you can review the code of the func I posted above, it uses the `s.api.GasEstimateMessageGas `, at that time, the msg's from is worker address, and it will get an `actor balance less than needed` error, then the func just return and will not check whether any other address is available

Got it!

**Comment by PatrickDeuse** [comment]
are you sure the error is coming from where you say?

does this 
https://github.com/filecoin-project/lotus/blob/7e55c44d4e09a058782e07cf172c96fb5546b172/node/impl/full/gas.go#L284
return a "actor balance less than needed"? (called here https://github.com/filecoin-project/lotus/blob/7e55c44d4e09a058782e07cf172c96fb5546b172/storage/wdpost_run.go#L769)

the address is selected here:

https://github.com/filecoin-project/lotus/blob/7e55c44d4e09a058782e07cf172c96fb5546b172/storage/wdpost_run.go#L802

**Comment by GFZRZK** [comment]
Yes, `gas estimation` will check the balance. It's run on a vm. That will check the balance.

**Comment by PatrickDeuse** [comment]
you're right. this should actually let the post fail more often. 

and in addition to the worker set as sender the gas estimate function possibly could use another tipset than the actual select address call.

**Comment by GFZRZK** [comment]
@jennijuju  I think there's no need to input something anymore, things are clear, and just redesign the code here is ok

**Comment by jennijuju** [comment]
@GFZRZK Glad you resolved it! did you create a pr or would you mind writing up how to redesign the code here so someone can pick it up? Thanks!

**Comment by gradoj** [comment]
winpost unexpectedly fell back to owner address causing failed winpost. 

```
021-02-15T17:35:30.503-0700	WARN	storageminer	storage/wdpost_run.go:225	Checked sectors	{"checked": 36, "good": 36}
2021-02-15T17:35:30.729-0700	ERROR	storageminer	storage/wdpost_run.go:447	checking sector recoveries: pushing message to mpool: failed to sign message: LedgerHID device (idx 0) not found
2021-02-15T18:35:30.369-0700	WARN	storageminer	storage/wdpost_run.go:225	Checked sectors	{"checked": 2, "good": 2}
2021-02-15T18:35:30.377-0700	INFO	storageminer	storage/wdpost_run.go:578	running window post	{"chain-random": "/R8ewH5ZF5KO/Wxl3lHAbDSh2Dcv/ln7Mz2CotjF27Q=", "deadline": {"CurrentEpoch":504431,"PeriodStart":504451,"Index":0,"Open":504451,"Close":504511,"Challenge":504431,"FaultCutoff":504381,"WPoStPeriodDeadlines":48,"WPoStProvingPeriod":2880,"WPoStChallengeWindow":60,"WPoStChallengeLookback":20,"FaultDeclarationCutoff":70}, "height": "504431", "skipped": 0}
2021-02-15T18:56:35.034-0700	INFO	storageminer	storage/wdpost_run.go:594	computing window post	{"batch": 0, "elapsed": 1264.657610139}
2021-02-15T18:56:36.529-0700	ERROR	storageminer	storage/wdpost_run.go:184	submit window post failed: pushing message to mpool:
    github.com/filecoin-project/lotus/storage.(*WindowPoStScheduler).submitPost
        /home/user/lotus/storage/wdpost_run.go:747
  - failed to sign message: LedgerHID device (idx 0) not found
```


```
$ lotus-miner actor control list
name       ID        key           use        balance                   
owner      f0151150  f1wdo7pku...             2.392644669478354899 FIL  
worker     f0152193  f3td2zmbn...  other      8.794460016429414735 FIL  
control-0  f0152123  f1kndg3qp...  precommit  5.17175832429380499 FIL   
control-1  f0151227  f1orue2ye...  commit     1.946931472293000624 FIL  
```
